3_harmonize/src/format_columns.R

```diff
@@ -0,0 +1,130 @@
+                        "remove the following variables from drop_vars: \n%s\n"),
+                 drop_vars_to_keep))
+  }
+ 
+  # Coerce variables to numeric, but first, retain original values in a new column
+  wqp_data_out <- wqp_data %>%
+    mutate(across(all_of(vars_to_numeric), ~., .names = "{col}_original"),
+           across(all_of(vars_to_numeric), as.numeric)) %>%
+    # format column order so that "original" cols are located next to formatted 
+    # vars_to_numeric cols
+    select(c("OrganizationIdentifier", "MonitoringLocationIdentifier",
+             "ActivityStartDate", "ActivityStartTime.Time",
+             "ActivityStartTime.TimeZoneCode", "CharacteristicName"), 
+           any_of(ends_with("_original")),
+           any_of(vars_to_numeric),
+           everything()) %>%
+    # suppress warnings including "NAs introduced by coercion" that will
+    # appear if values within a column in `vars_to_numeric` contain text
+    # and so cannot be parsed to a numeric value.
+    suppressWarnings() %>%
+    # drop any undesired columns 
+    select(-c(any_of(drop_vars)))
+    
+  # Save a file containing which values were replaced with NA when column
+  # was converted from character to numeric  
+  save_NA_vals(wqp_data, save_dir, vars_to_numeric)
+  
+  return(wqp_data_out)
+}
+
+
+#' @title Record values that have been replaced with NA
+#' 
+#' @description
+#' Function to save an output file that contains any values that were replaced
+#' with NA when a column was converted from class character to class numeric. 
+#' Values were replaced with NA if the original entry could not be parsed as 
+#' a numeric value.
+#' 
+#' @param wqp_data data frame containing the data downloaded from the WQP, 
+#' where each row represents a unique data record.
+#' @param save_dir character string indicating the directory where an output
+#' file should be saved. The output file is intended to inform the user of any
+#' WQP entries in columns `vars_to_numeric` that were not able to be parsed to
+#' numeric values and so were replaced with NA.
+#' @param vars_to_numeric character string indicating which column(s) should be 
+#' coerced to class numeric. 
+#' 
+#' @returns 
+#' Returns a character string indicating the file path of the saved output file.
+#' The saved .txt file includes a section for each column in `vars_to_numeric` 
+#' that contains the original WQP entries for that column that were replaced with 
+#' NA when the column was converted from class character to class numeric.
+#' 
+save_NA_vals <- function(wqp_data, save_dir, vars_to_numeric){
+
+  # Find any values from the original columns that were replaced with NA
+  values_replaced_with_na <- lapply(vars_to_numeric, function(x){
+    
+    vars_vals_na <- wqp_data %>%
+      filter(!is.na(.data[[x]]),
+             is.na(as.numeric(.data[[x]]))) %>%
+      suppressWarnings() %>%
+      pull(.data[[x]]) %>% 
```

Comment:
Yeah, I think either implementation will work. `.data[[x]]` is a 'pronoun' object implemented in `rlang` that can be used to access data variables. I use it because it _seems_ to be a recommended practice in dplyr programming (a la this [vignette](https://cran.r-project.org/web/packages/dplyr/vignettes/programming.html)), but I'm honestly not really sure if there are other differences.

Reply:
Wow! Sad that it is potentially the recommended practice since it appears in the dplyr vignette but I have never noticed! Thanks for sharing